### PR TITLE
[luci] quantize convolution's weight to layer-wise uint8

### PR DESCRIPTION
**Pull request by meejeong**
- apply to quantize convolution's only weights and bias to layer-wise 
- datatype : uint8
- asymmetric quantization

ONE-DCO-1.0-Signed-off-by: mia park <mia.park@samsung.com>

Related to: #696, https://github.sec.samsung.net/STAR/nnapps/issues/379

### CODE REVIEW

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp [hunk 1]
```diff
@@ -111,29 +111,31 @@ std::pair<loco::Node *, loco::Node *> get_input_weight_of_bias(CircleNode *node)
...
+void asym_quant_bias(CircleConst *node, float input_scale, float weight_scale,
+                     float &scaling_factor, int64_t &zp)
 {
   float scale = input_scale * weight_scale;
   const float scaling_factor_inv = (scale == 0) ? 0 : 1.0 / scale;
 
   uint32_t size = node->size<loco::DataType::FLOAT32>();
   std::vector<int32_t> quantized_values(size);
-  for (int i = 0; i < static_cast<int32_t>(size); ++i)
+  for (uint32_t i = 0; i < size; ++i)
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:122]: This change doesn't seem to be related with this PR

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp [hunk 2]
```diff
@@ -111,29 +111,31 @@ std::pair<loco::Node *, loco::Node *> get_input_weight_of_bias(CircleNode *node)
   return std::make_pair(nullptr, nullptr);
 }
 
-void quant_bias(CircleConst *node, float input_scale, float weight_scale, float *scaling_factor,
-                int64_t *zp)
+void asym_quant_bias(CircleConst *node, float input_scale, float weight_scale,
+                     float &scaling_factor, int64_t &zp)
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:115]: changing ptr to ref is also doesn't seem to be related with this PR

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp [hunk 3]
```diff
@@ -111,29 +111,31 @@ std::pair<loco::Node *, loco::Node *> get_input_weight_of_bias(CircleNode *node)
...
         static_cast<int32_t>(std::round(node->at<loco::DataType::FLOAT32>(i) * scaling_factor_inv));
   }
 
   node->dtype(loco::DataType::S32);      // change the type of tensor
   node->size<loco::DataType::S32>(size); // resize tensor
-  const int32_t kScale = std::numeric_limits<int32_t>::max();
-  for (int i = 0; i < static_cast<int32_t>(size); ++i)
+  const int32_t kMinScale = std::numeric_limits<int32_t>::lowest();
+  const int32_t kMaxScale = std::numeric_limits<int32_t>::max();
+  for (uint32_t i = 0; i < size; ++i)
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:132]: ditto

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp [hunk 4]
```diff
@@ -147,37 +149,24 @@ bool is_quantized(const CircleNode *node)
...
-    *zp = 0;
-    return;
-  }
 
-  compute_asym_scale_zp(min, max, scaling_factor, zp);
-  const float scaling_factor_inv = 1.0 / *scaling_factor;
+  const float scaling_factor_inv = 1.0 / scaling_factor;
   std::vector<int32_t> quantized_values(size);
-  for (int i = 0; i < static_cast<int32_t>(size); ++i)
+  for (uint32_t i = 0; i < size; ++i)
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:161]: ditto

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp [hunk 5]
```diff
@@ -147,37 +149,24 @@ bool is_quantized(const CircleNode *node)
...
-    quantized_values[i] = static_cast<int32_t>(
-        std::round(*zp + node->at<loco::DataType::FLOAT32>(i) * scaling_factor_inv));
+    auto data = node->at<loco::DataType::FLOAT32>(i);
+    quantized_values[i] = static_cast<int32_t>(std::round((data - min) * scaling_factor_inv));
   }
 
   node->dtype(loco::DataType::U8);      // change the type of tensor
   node->size<loco::DataType::U8>(size); // resize tensor
-  for (int i = 0; i < static_cast<int32_t>(size); ++i)
+  for (uint32_t i = 0; i < size; ++i)
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:169]: ditto

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp [hunk 6]
```diff
@@ -147,37 +149,24 @@ bool is_quantized(const CircleNode *node)
          node->dtype() == loco::DataType::S32;  // bias
 }
 
-void asym_wquant_with_minmax(CircleConst *node, float min, float max, float *scaling_factor,
-                             int64_t *zp)
+void asym_wquant_with_minmax(CircleConst *node, float min, float scaling_factor)
```
Comment (jinevening) [inline compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:152]: It would be better to change the function name.

Maybe something like `asym_wquant_with_min_scale` or simply `asym_wquant`.

I'd preferred the second one.

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp [hunk 7]
```diff
@@ -345,24 +343,22 @@ struct QuantizeWeights final : public luci::CircleNodeMutableVisitor<bool>
...
         {
-          auto data = circle_const->at<loco::DataType::FLOAT32>(i);
-          min = data < min ? data : min;
-          max = data > max ? data : max;
+          // will be implemented.
+        }
+        // Find min/max per layer-wise
+        else
+        {
+          // Quantize using recorded min/max
```
Comment (jinevening) [inline compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:354]: recorded min/max -> recorded quantparam

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp (hunk 6 above)
Comment (meejeong) [reply compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:152]: I also agree the second one :-) I will apply it.

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp [hunk 8]
```diff
@@ -292,22 +283,28 @@ struct QuantizeBias final : public luci::CircleNodeMutableVisitor<bool>
...
+    {
+      assert(input->quantparam()->scale.size() == 1); // Only support per-layer quant
+      auto input_scale = input->quantparam()->scale[0];
+
+      assert(weight->quantparam()->scale.size() == 1); // Only support per-layer quant
+      auto weight_scale = weight->quantparam()->scale[0];
+
+      auto circle_const = loco::must_cast<luci::CircleConst *>(node);
+      float scaling_factor;
+      int64_t zp;
```
Comment (mhs4670go) [inline compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:300]: Please initialize it to prevent warning in static analysis

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp [hunk 9]
```diff
@@ -345,24 +343,22 @@ struct QuantizeWeights final : public luci::CircleNodeMutableVisitor<bool>
...
-          auto data = circle_const->at<loco::DataType::FLOAT32>(i);
-          min = data < min ? data : min;
-          max = data > max ? data : max;
+          // will be implemented.
+        }
+        // Find min/max per layer-wise
+        else
+        {
+          // Quantize using recorded quantparam
+          auto quantparam = circle_node->quantparam();
```
Comment (mhs4670go) [inline compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:355]: maybe it would be better to assert `quantparam != nullptr`.

Comment (meejeong) [reply compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:355]: I applied it :-)

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp [hunk 10]
```diff
@@ -270,12 +259,14 @@ struct QuantizeActivation final : public luci::CircleNodeMutableVisitor<bool>
 
 struct QuantizeBias final : public luci::CircleNodeMutableVisitor<bool>
 {
-  QuantizeBias(loco::DataType input, loco::DataType output) : input_type(input), output_type(output)
+  QuantizeBias(loco::DataType input, loco::DataType output, QuantizationGranularity granularity)
+      : input_type(input), output_type(output), granularity(granularity)
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:263]: `granularity(granularity)` is same variable. does this work correctly?

Comment (meejeong) [reply compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:263]: when I execute this code, it works properly. However, I can change it if you think it is confused.

Comment (seanshpark) [reply compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:263]: Ah.. thanks, let's change it like `granularity(gr)`  :)

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp [hunk 11]
```diff
@@ -318,13 +315,14 @@ struct QuantizeBias final : public luci::CircleNodeMutableVisitor<bool>
  */
 struct QuantizeWeights final : public luci::CircleNodeMutableVisitor<bool>
 {
-  QuantizeWeights(loco::DataType input, loco::DataType output)
-      : input_type(input), output_type(output)
+  QuantizeWeights(loco::DataType input, loco::DataType output, QuantizationGranularity granularity)
+      : input_type(input), output_type(output), granularity(granularity)
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:319]: same here :)

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp (hunk 10 above)
Comment (mhs4670go) [reply compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:263]: AFAIK, It works well

Comment (mhs4670go) [reply compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:263]: AFAIK, it works well:)

Comment (s-barannikov) [reply compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp]: This indeed works well, but we usually use underscore prefix for private variables, e.g. "_granularity".
`QuantizeBias` should be a class rather than a struct, and the variables should be private.
It's just a note, please proceed without my approval if others find changes OK.
